forem/forem
# Manage Badges in /admin/badges

## What type of PR is this? (check all applicable)


## Description
Overall, this PR adds additional functionality to badge management and badge awarding via `/admin/badge_achievements` and specs to support the additions/changes that have been made. These additions and changes give Admins the ability to `create`, `update`, and "award" badges, as well as see an overview of all of the current badges via the redesigned `/admin/badges` index page. Furthermore, this PR addresses the verbiage around awarding badges and fixes a bug with usernames not being downcased when awarding badges to a user.

This PR adds the following:
- The ability to create new badges via new/create controller actions and the necessary view and routes to make this happen.
- The ability to edit/update existing badges via new controller actions and the necessary view and routes to make this happen.
- The ability to view all badges and click on individual badges in order to edit/update them via a redesigned index page.
- Appropriate methods in `Admin::Internal::Badges` and routes for the changes made.
- A button, "Award Badges", that sends Admins to the badge award form, where they can award badges to users.
- A button, "Create Badges", that sends Admins to the badge creation form, where they can create new badges.
- Updates the routes for awarding badges from `/admin/badges/award_badges` to `/admin/badge_achievements/award_badges`
- Reverts changes made to redirecting from `/admin/badges` to `/admin/badge_achievements` (after a discussion, we decided that only awarding badges should live under the `/badge_achievements` route and every other action should remain living under `/admin/badges`).
- Updates to the language used in the warning banner on the badge award form and above the usernames field on the form.
- `.downcase` to usernames when awarding badges to ensure that they are lowercase and do not silently fail.
- Additional specs around case-sensitive names when awarding badges, the new badge URL and its redirect, updating badges with the new `edit` form, and creating badges with the `new` form.

## Related Tickets & Documents
Related to PR #9643 

## QA Instructions, Screenshots, Recordings

- Navigate to `/admin/badges` to observe the redesigned `index` page for badge management:
<img width="1377" alt="Screen Shot 2020-08-11 at 1 58 56 PM" src="https://user-images.githubusercontent.com/32834804/89943055-d6200e80-dbda-11ea-8b97-7e8c702520ec.png">

- Click on an individual badge's `title` to navigate to the badge's `edit` page:
<img width="1377" alt="Screen Shot 2020-08-11 at 2 29 20 PM" src="https://user-images.githubusercontent.com/32834804/89946010-7bd57c80-dbdf-11ea-859f-d522ea2398e9.png">

- Once on the `edit` page, update the badge by changing its `title`, `slug`, `description`, and/or `badge_image` to ensure that everything works properly (prior to updating everything, test that the "Back to All Badges" button works!):
<img width="1377" alt="Screen Shot 2020-08-11 at 2 31 15 PM" src="https://user-images.githubusercontent.com/32834804/89946025-82fc8a80-dbdf-11ea-98a4-8e8986d86d84.png">

- After updating the badge, you will be redirected back to the `index` page. You should see your newly updated badge appear in the table, as well as a confirmation that the badge was updated:
<img width="1377" alt="Screen Shot 2020-08-11 at 2 31 35 PM" src="https://user-images.githubusercontent.com/32834804/89946046-8b54c580-dbdf-11ea-8283-01aaf2292b82.png">

- Once you are redirected back to the `index` page, click the "Create Badge" button to navigate to the badge creation form:
<img width="1377" alt="Screen Shot 2020-08-11 at 2 36 06 PM" src="https://user-images.githubusercontent.com/32834804/89946542-54cb7a80-dbe0-11ea-8706-879ebf7fcca6.png">

- Fill out the form and click the "Create Badge" button to successfully create a new badge:
<img width="1377" alt="Screen Shot 2020-08-11 at 2 37 45 PM" src="https://user-images.githubusercontent.com/32834804/89946565-5eed7900-dbe0-11ea-9ee9-400f94f5c839.png">

- After creating the badge, you will be redirected back to the `index` page. You should see your newly created badge appear in the table, as well as a confirmation that the badge was created:
<img width="1377" alt="Screen Shot 2020-08-11 at 2 37 58 PM" src="https://user-images.githubusercontent.com/32834804/89946582-67de4a80-dbe0-11ea-8b9b-e4e1cf16a330.png">

- Once you are back on the `index` page, click the "Award Badge" button to navigate to the badge awarding form:
<img width="1377" alt="Screen Shot 2020-08-11 at 2 44 13 PM" src="https://user-images.githubusercontent.com/32834804/89947174-4a5db080-dbe1-11ea-8863-6dc5c707e6f9.png">

- Select a badge and fill out the rest of the form and to award a badge to yourself and/or another user:
<img width="1377" alt="Screen Shot 2020-08-11 at 2 44 56 PM" src="https://user-images.githubusercontent.com/32834804/89947209-53e71880-dbe1-11ea-9ecb-bca7ec0133e8.png">

- Once you have clicked "Award Badges", you will be redirected to the `index` page where you should see a confirmation message:
<img width="1377" alt="Screen Shot 2020-08-11 at 2 39 53 PM" src="https://user-images.githubusercontent.com/32834804/89946723-95c38f00-dbe0-11ea-8037-06ef87c0748a.png">

- Navigate to your notifications to ensure that you received a notification for the awarded badge:
<img width="1341" alt="Screen Shot 2020-08-11 at 11 15 51 AM" src="https://user-images.githubusercontent.com/32834804/89927855-30fa3b80-dbc4-11ea-82c6-444870e5b314.png">

- Navigate to the profiles of the users you awarded badges to and ensure that the badges were awarded:
<img width="1377" alt="Screen Shot 2020-08-11 at 11 59 00 AM" src="https://user-images.githubusercontent.com/32834804/89931934-217df100-dbca-11ea-91ba-717d7fa0320a.png">
<img width="1377" alt="Screen Shot 2020-08-11 at 2 55 04 PM" src="https://user-images.githubusercontent.com/32834804/89948067-ae34a900-dbe2-11ea-8428-35bebb5b03d4.png">

- Click on the newly awarded badge to ensure that it renders as expected with all of the badge information you provided:
<img width="1023" alt="Screen Shot 2020-08-11 at 11 57 41 AM" src="https://user-images.githubusercontent.com/32834804/89931904-188d1f80-dbca-11ea-9888-02fff467adc0.png">

- Finally, navigate to one of your badges using `/badge/` + the `slug` for the badge (for example, `/badge/money-sloth-award`) it to ensure that navigating this way still works as well:
<img width="1341" alt="Screen Shot 2020-08-11 at 11 52 34 AM" src="https://user-images.githubusercontent.com/32834804/89931406-55a4e200-dbc9-11ea-8c75-ee7dcf3dbb7d.png">

- Try to break things! 🛠️ 

## Added tests?

- [x] yes, in addition to updating existing specs

## Added to documentation?

- [x] no documentation needed... I think? Any ideas on whether or not these addition/changes warrant additions/changes to documentation?

## [optional] Are there any post-deployment tasks we need to perform?
None that I can think of!

## [optional] What gif best describes this PR or how it makes you feel?

![John Legend Award](https://media.giphy.com/media/3oEduPq5xmKym6on16/giphy.gif)
